### PR TITLE
Jetty 12.0.x xmlconfig prefers interfaces

### LINKED ISSUE #11895: XmlConfiguration should prefer exact class match rather than interface match
**jetty-12**

When matching xml to class methods to satisfy a `<Call>` clause, the code currently prefers matching an argument that is an interface rather than an exact class type match. 
In other words, given the following java code:

```java

    public class Example
    {
      interfact TestInterface {}
      public class TestImpl implements TestInterface {}
 
      public static void callStaticWithVarArgs(TestInterface i, String... strings)
      {
      }

      public static void callStaticWithVarArgs(TestImpl i, String... strings)
      {
      }
  }
```

And given the following xml:

```xml
   <Call class="Example" name="callStaticWithVarArgs">
    <Arg>
      <New class="Example$TestImpl"/>
    </Arg>
    <Arg>
      <Array type="java.lang.String">
        <Item>A</Item>
        <Item>B</Item>
        <Item>C</Item>
      </Array>
    </Arg>
  </Call>
```

The method `callStaticWithVarArgs(TestInterface i, String... strings)` is ranked higher than the alternative method - which is actually an exact match - and thus selected.

### PR DESCRIPTION
**Pull request by janbartel**
Fixes #11895.

Tentative fix to improve ordering of method matches so that:

* methods with fewer args are preferred
* methods with primitive args are preferred
* methods with non-varargs are preferred
* methods with class args (that are not the Object.class) are preferred
* methods with interface args or class args preference the most derived type

### CODE REVIEW

**Comment by joakime** [comment]
@janbartel rebase against jetty-12.1.x

**Comment by gregw** [comment]
I think this is a good change for 12.1,  I'm not so worried by this now (cue disaster!)

**Comment by gregw** [comment]
@janbartel can you rebase this to 12.1, I think it is a good change to make.

**Comment by janbartel** [comment]
Rebase done.

File: jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 1]
```diff
@@ -110,6 +110,32 @@ public class XmlConfiguration
...
+     * an interface type. Where both are the Object type, or both are an interface they
+     * are considered equal.
+     *
+     * @param t1 the first class to compare
+     * @param t2 the second class to compare
+     * @return 1 if t1 is a non Object class or an interface type, return 0 if both are
+     * interfaces or the Object class, return -1 if only t2 is the Object class or an
+     * interface.
+     */
+    public static int compare(Class<?> t1, Class<?> t2)
```
Comment (gregw) [inline jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: private?

File: jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 2]
```diff
@@ -1007,6 +1033,7 @@ private Object call(Class<?> oClass, String methodName, Object obj, Args args) t
 
                 try
                 {
+                    System.err.println("Invoking: " + method);
```
Comment (gregw) [inline jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: remove

File: jetty-core/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java [hunk 3]
```diff
@@ -2165,15 +2178,21 @@ public void testFooExecutableComparator()
...
+        }
+
         String[] expectedOrder = {
-            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo()", // favor fewer args
-            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(int)", // favor primitives over non-primitives
-            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.time.temporal.Temporal)", // favor over Instant
-            "public int org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.lang.String)",
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo()", // favour fewer args
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(int)", // favour primitives over non-primitives
+            "public int org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.lang.String)", //favour classes over interfaces
```
Comment (gregw) [inline jetty-core/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java:2222]: Can you add a `foo(CharSequence)` to the example, as `CharSequence` is an interface implemented by `String`

**Review by gregw — CHANGES_REQUESTED** [review]

**Review by gregw — APPROVED** [review]
LGTM

**Pull request merged**